MaterialDesignThemes.Wpf/PopupBox.cs

```diff
@@ -439,21 +454,73 @@ public override void OnApplyTemplate()
             VisualStateManager.GoToState(this, IsPopupOpen ? PopupIsOpenStateName : PopupIsClosedStateName, false);
         }
 
-        protected override void OnIsKeyboardFocusWithinChanged(DependencyPropertyChangedEventArgs e)
+        //protected override void OnIsKeyboardFocusWithinChanged(DependencyPropertyChangedEventArgs e)
+        //{
+        //    base.OnIsKeyboardFocusWithinChanged(e);
+
+        //    if (IsPopupOpen && !IsKeyboardFocusWithin && !StaysOpen)
+        //    {
+        //        Close();
+        //    }
+        //}
+
+        bool supressfirstElementGotFocus = true;
+        protected override void OnGotKeyboardFocus(KeyboardFocusChangedEventArgs e)
         {
-            base.OnIsKeyboardFocusWithinChanged(e);
+            base.OnGotKeyboardFocus(e);
+            if (supressfirstElementGotFocus == false) { return; }
 
-            if (IsPopupOpen && !IsKeyboardFocusWithin && !StaysOpen)
+            if (!IsPopupOpen && PopupMode == PopupBoxPopupMode.ClickOrFocusWithin && IsKeyboardFocusWithin)
             {
-                Close();
+                SetCurrentValue(IsPopupOpenProperty, true);
+                if (_popupContentControl != null)
+                {
+                    _popupContentControl.MoveFocus(new TraversalRequest(FocusNavigationDirection.First));
+
+                    //getting the first element, and ensure that when it gets focus again, move the focus outside of the Popup
+                    //to avoid unwanted tab stop cycling only inside it.
+                    IInputElement firstelement = Keyboard.FocusedElement;
+                    if (firstelement == null) { return; }
+                    firstelement.LostKeyboardFocus += InnerFirstElement_LostKeyboardFocus;
+
+                    firstelement.GotKeyboardFocus += (s, e) =>
```

Comment:
I think you might be registering another event handlers every time `OnGotKeyboardFocus` runs?

Reply:
You are right. I'll resolve this. Thanks.